opm/simulators/linalg/bda/Matrix.hpp

```diff
@@ -17,41 +17,83 @@
   along with OPM.  If not, see <http://www.gnu.org/licenses/>.
 */
 
-#ifndef FPGA_MATRIX_HEADER_INCLUDED
-#define FPGA_MATRIX_HEADER_INCLUDED
+#ifndef MATRIX_HEADER_INCLUDED
+#define MATRIX_HEADER_INCLUDED
 
 #include <vector>
 
+#include <opm/simulators/linalg/bda/opencl.hpp>
+#include <opm/simulators/linalg/bda/BlockedMatrix.hpp>
+
 namespace Opm
 {
 namespace Accelerator
 {
 
+class Matrix;
+
+/// This struct resembles a csr matrix, only doubles are supported
+/// The matrix data is stored in OpenCL Buffers
+class OpenclMatrix {
+public:
+
+    OpenclMatrix(cl::Context *context, int Nb_, int Mb_, int nnzbs_, unsigned int block_size_)
+    : Nb(Nb_),
+      Mb(Mb_),
+      nnzbs(nnzbs_),
+      block_size(block_size_)
+    {
+        nnzValues = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(double) * block_size * block_size * nnzbs);
+        colIndices = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(int) * nnzbs);
+        rowPointers = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(int) * (Nb + 1));
+    }
+
+    void upload(cl::CommandQueue *queue, double *vals, int *cols, int *rows);
+    void upload(cl::CommandQueue *queue, Matrix *matrix);
+    void upload(cl::CommandQueue *queue, BlockedMatrix *matrix);
+
+    cl::Buffer nnzValues;
+    cl::Buffer colIndices;
+    cl::Buffer rowPointers;
+    int Nb, Mb;
+    int nnzbs;
+    unsigned int block_size;
+};
+
+
 /// This struct resembles a csr matrix, only doubles are supported
 /// The data is stored in contiguous memory, such that they can be copied to a device in one transfer.
 class Matrix {
 
 public:
 
-    /// Allocate Matrix and data arrays with given sizes
+    /// Allocate square Matrix and data arrays with given sizes
     /// \param[in] N               number of rows
     /// \param[in] nnzs            number of nonzeros
     Matrix(int N_, int nnzs_)
-    : nnzValues(new double[nnzs_]),
-      colIndices(new int[nnzs_]),
-      rowPointers(new int[N_+1]),
-      N(N_),
+    : N(N_),
       nnzs(nnzs_)
-    {}
-
-    /// All constructors allocate new memory, so always delete here
-    ~Matrix(){
-        delete[] nnzValues;
-        delete[] colIndices;
-        delete[] rowPointers;
+    {
+        nnzValues.resize(nnzs);
+        colIndices.resize(nnzs);
+        rowPointers.resize(N+1);
     }
 
+    /// Allocate rectangular Matrix and data arrays with given sizes
+    /// \param[in] N               number of rows
+    /// \param[in] M               number of columns
+    /// \param[in] nnzs            number of nonzeros
+    Matrix(int N_, int M_, int nnzs_)
+    : N(N_),
+      M(M_),
+      nnzs(nnzs_)
```

Comment:
I think you can call other constructors here and save some code:
```
Matrix(int N_, int M_, int nnzs_)
: Matrix(N_, nnzs_), M(M_)
{}
```

Reply:
845485ae99